### PR TITLE
[ttl] Add support matmul + elementwise fusion (fixes #476)

### LINKED ISSUE #476: [external bug report][from claude] Elementwise ops silently dropped when fused with matmul in a single store
## Summary

When a matmul result is used in elementwise operations within a single `.store()` call, the elementwise operations are silently dropped. The kernel compiles and runs without error, but produces incorrect results.

```python
# BUG: produces just `a @ b`, the scale and bias are silently dropped
o.store(sc * (a @ b) + bi)

# WORKAROUND: split matmul into its own store
mm.store(a @ b)
# ... then in a separate with block:
o.store(sc * mm + bi)
```

## Observed behavior

`o.store(sc * (a @ b) + bi)` produces the same output as `o.store(a @ b)`. The `sc *` and `+ bi` have no effect.

HW output:
```
ref (scale*mm+bias)[0,:3]: [11.875, 6.3125, 8.9375]
raw matmul[0,:3]:          [3.765625, -7.375, -2.0625]
fused result[0,:3]:        [3.765625, -7.375, -2.0625]   <- matches raw matmul, NOT ref
split result[0,:3]:        [11.875, 6.3125, 9.0]         <- correct
```

## Expected behavior

`o.store(sc * (a @ b) + bi)` should produce the same result as the two-step version, or raise an error if this fusion is not supported.

## Impact

This is a silent correctness bug. We hit this while writing an attention backward kernel -- the expression `s.store(scale * (Q @ K_T) + mask)` silently dropped both the attention scale and the causal mask, producing wildly wrong gradients (~5e10x error) with no compilation warning.

## Reproducer

Minimal self-contained reproducer comparing fused vs split kernels:

```python
"""Reproducer: elementwise ops fused with matmul in a single store are silently dropped."""
import torch
import ttl
import ttnn

TILE = 32

def to_ttnn(t, device):
    return ttnn.from_torch(t, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
                           device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG)

def to_ttnn_l1(t, device):
    return ttnn.from_torch(t, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
                           device=device, memory_config=ttnn.L1_MEMORY_CONFIG)

# --- Buggy: fused scalar * matmul + bias in one store ---
@ttl.operation(grid=(1, 1))
def fused_kernel(A, B, scale_tile, bias_tile, out):
    a_dfb = ttl.make_dataflow_buffer_like(A, shape=(1, 1), buffer_factor=2)
    b_dfb = ttl.make_dataflow_buffer_like(B, shape=(1, 1), buffer_factor=2)
    sc_dfb = ttl.make_dataflow_buffer_like(scale_tile, shape=(1, 1), buffer_factor=1)
    bias_dfb = ttl.make_dataflow_buffer_like(bias_tile, shape=(1, 1), buffer_factor=2)
    out_dfb = ttl.make_dataflow_buffer_like(out, shape=(1, 1), buffer_factor=2)

    @ttl.compute()
    def compute():
        with sc_dfb.wait() as sc:
            with a_dfb.wait() as a, b_dfb.wait() as b, bias_dfb.wait() as bi:
                with out_dfb.reserve() as o:
                    o.store(sc * (a @ b) + bi)  # BUG: scale and bias silently dropped

    @ttl.datamovement()
    def dm_read():
        with sc_dfb.reserve() as blk:
            tx = ttl.copy(scale_tile[0, 0], blk); tx.wait()
        with a_dfb.reserve() as blk:
            tx = ttl.copy(A[0, 0], blk); tx.wait()
        with b_dfb.reserve() as blk:
            tx = ttl.copy(B[0, 0], blk); tx.wait()
        with bias_dfb.reserve() as blk:
            tx = ttl.copy(bias_tile[0, 0], blk); tx.wait()

    @ttl.datamovement()
    def dm_write():
        with out_dfb.wait() as blk:
            tx = ttl.copy(blk, out[0, 0]); tx.wait()

# --- Correct: matmul in its own store, then elementwise ---
@ttl.operation(grid=(1, 1))
def split_kernel(A, B, scale_tile, bias_tile, out):
    a_dfb = ttl.make_dataflow_buffer_like(A, shape=(1, 1), buffer_factor=2)
    b_dfb = ttl.make_dataflow_buffer_like(B, shape=(1, 1), buffer_factor=2)
    sc_dfb = ttl.make_dataflow_buffer_like(scale_tile, shape=(1, 1), buffer_factor=1)
    bias_dfb = ttl.make_dataflow_buffer_like(bias_tile, shape=(1, 1), buffer_factor=2)
    mm_dfb = ttl.make_dataflow_buffer_like(out, shape=(1, 1), buffer_factor=2)
    out_dfb = ttl.make_dataflow_buffer_like(out, shape=(1, 1), buffer_factor=2)

    @ttl.compute()
    def compute():
        with sc_dfb.wait() as sc:
            with a_dfb.wait() as a, b_dfb.wait() as b, bias_dfb.wait() as bi:
                with mm_dfb.reserve() as mm:
                    mm.store(a @ b)
                with mm_dfb.wait() as mm:
                    with out_dfb.reserve() as o:
                        o.store(sc * mm + bi)

    @ttl.datamovement()
    def dm_read():
        with sc_dfb.reserve() as blk:
            tx = ttl.copy(scale_tile[0, 0], blk); tx.wait()
        with a_dfb.reserve() as blk:
            tx = ttl.copy(A[0, 0], blk); tx.wait()
        with b_dfb.reserve() as blk:
            tx = ttl.copy(B[0, 0], blk); tx.wait()
        with bias_dfb.reserve() as blk:
            tx = ttl.copy(bias_tile[0, 0], blk); tx.wait()

    @ttl.datamovement()
    def dm_write():
        with out_dfb.wait() as blk:
            tx = ttl.copy(blk, out[0, 0]); tx.wait()

if __name__ == "__main__":
    device = ttnn.open_device(device_id=0)
    try:
        torch.manual_seed(42)
        A_pt = torch.randn(TILE, TILE, dtype=torch.bfloat16)
        B_pt = torch.randn(TILE, TILE, dtype=torch.bfloat16)
        scale_val = 0.5
        bias_pt = torch.full((TILE, TILE), 10.0, dtype=torch.bfloat16)

        ref = (scale_val * (A_pt.float() @ B_pt.float()) + bias_pt.float()).to(torch.bfloat16)
        raw_mm = (A_pt.float() @ B_pt.float()).to(torch.bfloat16)

        A_tt = to_ttnn(A_pt, device)
        B_tt = to_ttnn(B_pt, device)
        scale_tt = to_ttnn_l1(torch.full((TILE, TILE), scale_val, dtype=torch.bfloat16), device)
        bias_tt = to_ttnn_l1(bias_pt, device)

        out_fused = to_ttnn(torch.zeros(TILE, TILE, dtype=torch.bfloat16), device)
        fused_kernel(A_tt, B_tt, scale_tt, bias_tt, out_fused)
        result_fused = ttnn.to_torch(out_fused).reshape(TILE, TILE).float()

        out_split = to_ttnn(torch.zeros(TILE, TILE, dtype=torch.bfloat16), device)
        split_kernel(A_tt, B_tt, scale_tt, bias_tt, out_split)
        result_split = ttnn.to_torch(out_split).reshape(TILE, TILE).float()

        ref_f = ref.float()
        raw_f = raw_mm.float()

        print(f"ref[0,:3]:          {ref_f[0, :3].tolist()}")
        print(f"raw matmul[0,:3]:   {raw_f[0, :3].tolist()}")
        print(f"fused result[0,:3]: {result_fused[0, :3].tolist()}")
        print(f"split result[0,:3]: {result_split[0, :3].tolist()}")

        fused_vs_ref = (result_fused - ref_f).abs().max().item()
        fused_vs_raw = (result_fused - raw_f).abs().max().item()
        split_vs_ref = (result_split - ref_f).abs().max().item()
        print(f"\nfused vs ref:     {fused_vs_ref:.4f}")
        print(f"fused vs raw_mm:  {fused_vs_raw:.4f}")
        print(f"split vs ref:     {split_vs_ref:.4f}")
    finally:
        ttnn.close_device(device)
```

## Notes

- `o.store(old + a @ b)` (add + matmul) works correctly -- the issue is specifically when the matmul result is multiplied or has other ops chained before the add.
- Tested on BH hardware. Did not test on simulator (sim had separate issues with this kernel shape).

### PR DESCRIPTION
**Pull request by brnorris03**
Problem: When a matmul result is used in elementwise operations within a single `.store()` call, all the elementwise operations are silently dropped. The kernel compiles and runs without error, but produces incorrect results.

Cause: the separate `TTLLowerMatmulBlock` pass (which was just the initial implementation of matmul lowering) replaced the entire `ComputeOp` when a matmul was present, but only preserved unary in-place post-ops (filtered by `TTLInPlaceOpTrait`). Binary ops like `tile_mul` and `tile_add` were silently discarded.

Fix: Refactor matmul lowering to not just consume the `ttl.compute` containing the matmul.
- Merge matmul lowering into `TTLLowerToLoops` as a matmul-specific code path (`generateMatmulCompute`). When a `ComputeOp` contains `tile_matmul_block`, it emits a single `DstSectionOp` with the matmul call, all post-matmul ops (binary elementwise, `copy_tile`, etc.), and stores.
- Remove the separate `TTLLowerMatmulBlock` pass. The `use-block-matmul` option is preserved as a `TTLLowerToLoops` option for future `matmul_tile` support.
- Rename `TTLLowerMatmulBlock.cpp` to `LowerMatmulCompute.cpp`, added header at `include/ttlang/Dialect/TTL/Transforms/LowerMatmulCompute.h`.
- Support multi-output computes (same value stored to multiple DFBs). Partially addresses #509 (points 1 and 2; point 3 remains open).
- Deduplicate `applyIndexingMap` into `TTLOpsUtils.h`.
- Update pipeline documentation.

Closes #476

### CODE REVIEW

**Review by zoecarver — APPROVED** [review]

File: lib/Dialect/TTL/Transforms/LowerMatmulCompute.cpp [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+      continue;
+    }
+    bodyOpsToClone.push_back(&bodyOp);
+  }
+
+  // Determine the maximum DST index used by body ops.
+  int64_t maxBodyDstIdx = 0;
+  for (Operation *bodyOp : bodyOpsToClone) {
+    if (auto dstVal = getTileOpDstIndex(bodyOp)) {
+      if (auto constIdx = foldIndexToConstant(*dstVal)) {
```
Comment (zoecarver) [inline lib/Dialect/TTL/Transforms/LowerMatmulCompute.cpp]: maybe add assert/error if `foldIndexToConstant` returns null.
